instrumentation/mongo/mongo-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/mongo/v4_0/MongoClientInstrumentationModule.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.mongo.v4_0;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static net.bytebuddy.matcher.ElementMatchers.declaresMethod;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.event.CommandListener;
+import io.opentelemetry.javaagent.instrumentation.mongo.TracingCommandListener;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.lang.reflect.Modifier;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(InstrumentationModule.class)
+public class MongoClientInstrumentationModule extends InstrumentationModule {
+
+  public MongoClientInstrumentationModule() {
+    super("mongo", "mongo-4.0");
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return singletonList(new MongoClientSettingsBuilderInstrumentation());
+  }
+
+  private static final class MongoClientSettingsBuilderInstrumentation
+      implements TypeInstrumentation {
+    @Override
+    public ElementMatcher<TypeDescription> typeMatcher() {
+      return named("com.mongodb.MongoClientSettings$Builder")
+          .and(
+              declaresMethod(
+                  named("addCommandListener")
+                      .and(
+                          takesArguments(
+                              new TypeDescription.Latent(
+                                  "com.mongodb.event.CommandListener",
+                                  Modifier.PUBLIC,
+                                  null,
+                                  Collections.<TypeDescription.Generic>emptyList())))
+                      .and(isPublic())));
+    }
+
+    @Override
+    public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+      return singletonMap(
+          isMethod().and(isPublic()).and(named("build")).and(takesArguments(0)),
+          MongoClientInstrumentationModule.class.getName() + "$MongoClientAdvice");
+    }
+  }
+
+  public static class MongoClientAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void injectTraceListener(
+        @Advice.This MongoClientSettings.Builder builder,
+        @Advice.FieldValue("commandListeners") List<CommandListener> commandListeners) {
```

Comment:
Thanks for using the public method - for suppression we would want to instrument `addCommandListener` instead to avoid accessing this private field. However, we've generally been skipping handling duplicate registration in an effort to come with a more general pattern so you don't have to worry about it either if it's tedious.